torchtune/modules/multimodal_transformer.py

```diff
@@ -0,0 +1,206 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+import torch
+from torch import nn, Tensor
+
+from torchtune.modules import GroupedQueryAttention
+
+
+class TransformerSelfAttentionLayer(nn.Module):
+    """Transformer layer derived from the Llama2 model. Normalization is applied before the attention **and** FF layer.
+
+    Args:
+        attn (GroupedQueryAttention): Attention module.
+        mlp (nn.Module): Feed-forward module.
+        attn_norm (Optinoal[nn.Module]): Normalization to be applied before self-attention.
+        mlp_norm (Optional[nn.Module]): Normalization to be applied before the feed-forward layer.
+        attn_scale (Optinoal[nn.Module]): Module to scale self-attention output.
+        mlp_scale (Optional[nn.Module]): Module to scale the feed-forward output.
+    """
+
+    def __init__(
+        self,
+        attn: GroupedQueryAttention,
+        mlp: nn.Module,
+        *,
+        attn_norm: Optional[nn.Module] = None,
+        mlp_norm: Optional[nn.Module] = None,
+        attn_scale: Optional[nn.Module] = None,
+        mlp_scale: Optional[nn.Module] = None,
```

Comment:
Sorry to keep harping on this, but I still don't understand why we don't just provide separate versions of self-attention and MLP modules with scaling (see e.g. [here](https://gist.github.com/ebsmothers/7ed8a319b7566b76868299e245f5995b) for what the MLP would look like). Then we can provide different builders for `TransformerSelfAttentionLayer` with and without scaling and users don't have to try and figure out `attn_scale` and `mlp_scale` mean

Reply:
I think that both ways work. I like your idea, but i also think its convenient if the module already provides it. I guess that one benefits of this logic being in the transformer module is that we dont have to have 2x the implementation of every scale and MLP module, one gated and one that isnt. The con is that this module gets a bit bloated.